designs/auto-certs-renewal.md

```diff
@@ -0,0 +1,122 @@
+2. We can renew all cp’s certs besides apiserver-etcd-client by **recreating the cp machine.** As we can see from the following diagram that the apiserver-etcd-client is generated by Etcdadm Controller, so this cert cannot be renewed by just recreating the cp machine.
+3. No need to do auto renew certs in worker nodes.
+
+## Solution
+
+Etcd can be deployed in [two topologies.](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/ha-topology/) Certs renew mechanism is slightly different in those two scenarios.
+
+### Stacked etcd Nodes
+
+In Stacked etcd nodes topology, etcd is running inside the cp node. 
+
+We can renew all certificates by just redeploying cp nodes
+
+1. Expose [rolloutBefore.certificatesExpiryDays](https://cluster-api.sigs.k8s.io/tasks/certs/auto-rotate-certificates-in-kcp.html) field from KubeadmControlPlane in EKSA cluster spec. Add pass the field to KubeadmControlPlane
+
+### External etcd Nodes
+
+In external etcd nodes topology, etcd is running in a separate machine besides cp machine. In addition to roll out cp nodes, we also need to roll out etcd nodes, with the following changes on top of above:
+
+1. Add rolloutBefore.certificatesExpiryDays field to [EtcdadmCluster](https://github.com/aws/etcdadm-controller/blob/main/api/v1beta1/etcdadmcluster_types.go#L38), exposing and passing it from EKSA. The business logic of this field and reconciler should be:
+    1. apply machine annotation `machine.cluster.x-k8s.io/certificates-expiry` during etcd node creation.
+    2. check if roll out is needed. Refer https://github.com/kubernetes-sigs/cluster-api/blob/main/controlplane/kubeadm/internal/controllers/controller.go#L380.
+    3. if roll out is needed for any etcd node, upgrade all etcd nodes.
+
+However, not all certificates in cp node would be renewed even both etcd and cp nodes have changed, since apiserver-etcd-client key pair is created by etcdadm controller (refer the sequence diagram) and it always stays the same. We need to fix this bug by changing [etcdadm controller’s code](https://github.com/aws/etcdadm-controller/blob/main/controllers/certs.go#L81), it should update apiserver-etcd-client secret constantly in a short interval, such as every 24 hours. If we choose just update the secret during etcd nodes rolling out, then we have to invent mechanism to make sure the etcd nodes rolling out happens before cp nodes rolling out when it comes to renew certificates.
+
+### Cluster Yaml Spec Example
+
+```
+apiVersion: anywhere.eks.amazonaws.com/v1alpha1
+kind: Cluster
+spec:
+  certificatesConfiguration:
+    rolloutBeforeExpiry: 
+      // whether to rollout etcd and cp nodes before certs expiring
+      enabled: true/false (false by default, auto certs renew require opt-in) 
+      // allow adding finer control in the future
```

Comment:
What's the rationale for `rolloutBeforeExpiry` being a map vs a bool?

For example
```yaml
spec:
  certificates:
    rolloutBeforeExpiry: true
```

Also curious if, instead of a bool, we should use a time with [`time.Parse()`](https://pkg.go.dev/time#ParseDuration) semantics.

Reply:
It's a decision that we made with PM, the rationale is making it easier for the user to opt-in and flexible if we get feedbacks from users that they want to specify when to renew the certs